memtagabielf64/memtagabielf64.rst

```diff
@@ -0,0 +1,723 @@
+   holds the virtual address divided by the tag granule size (i.e.
+   ``descriptor.address = variable.address / 16``). Similarly, the size is
+   always granule-aligned, and the descriptor holds the size divided by the tag
+   granule size (i.e. ``descriptor.size = variable.size / 16``).
+
+2. Most global variables are laid out contiguously, and so the distance between
+   adjacent tagged global variables is short (often the next global is located
+   at ``previous_vaddr + previous_size``). Instead of encoding virtual
+   addresses, we sort the entire list of globals, ascending, by the virtual
+   address, and encode the distance (in granules) between the previous tagged
+   global's end address (``previous_vaddr + previous_size``) and itself. The
+   first global is offset by the number of granules from the zero address. For
+   example, if the first 32-byte tagged global is at ``0x100`` and the second is
+   at ``0x120``, then ``descriptor[0] = { .addr = 0x10, .size = 0x2 }`` and
+   ``descriptor[1] = { .addr = 0x0, .size = 0x2 }``. In order to minimise the
+   size of the descriptor section, an ideal linker will lay out all tagged
+   globals within a section contiguously.
+
+3. We encode the distance between globals, and the size of the global, as
+   ULEB128 integers.
+
+4. Often the ULEB128-encoded distance is less than a byte in size (there are
+   many globals separated by a distance of 0, 1, and 2 tags, and not many with a
+   distance larger than that). In addition, most of the globals are 1, 2, and 3
+   granules in size, but naively cost a full byte for the ULEB128-encoded value.
+   In order to optimise these common cases, the lower three bits of the distance
+   are reserved to encode the size. If the size of the global variable is less
+   than 8 (2\ :sup:`3`) tag granules, then the reserved bits of the distance
+   shall carry the size of the global. Otherwise, the reserved bits shall be set
+   to zero, and the size of the global shall be placed as an additional
+   ULEB128-encoded integer after the distance.
+
+5. A tagged global can't be zero tag granules in size. Thus, if an extra
+   ULEB128 integer is used for the size, the encoded value should be ``size -
+   1``.
+
+Pseudocode to generate the ``SHT_AARCH64_MEMTAG_GLOBALS_DYNAMIC`` section is
+presented below:
+
+.. code-block::
+
+   distance_reserved_bits = 3
+   sort(tagged_symbols) # by symbol.address, ascending
+
+   last_symbol_end = 0
+   for symbol in tagged_symbols:
+      distance = ((symbol.address - last_symbol_end) / tag_granule_size)
+                  << distance_reserved_bits
+      size = symbol.size / tag_granule_size
+      last_symbol_end = symbol.address + symbol.size
+
+      if size < (1 << distance_reserved_bits):
+         write_to_section(uleb128(distance | size))
+      else:
+         write_to_section(uleb128(distance))
+         write_to_section(uleb128(size - 1))
+
+.. raw:: pdf
+
+   PageBreak
+
+Relocation Operations
```

Comment:
I've done a quick check to see if any of this conflicts with the PAuthABI extension. As that uses new relocation codes https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#9auth-variant-dynamic-relocations it doesn't directly conflict.

It may be worth having a read of the PAuthABI to see if there is a scope for a signed pointer to a tagged location. While I don't think it is necessary to document all the details, it would be good to document what the expectation of the two extensions working together are. For example it could be that it is PAuthABI or Memtag and not both.

Reply:
Thanks for the hint, I'll keep this thread open as I investigate further.